src/V3HierBlock.cpp

```diff
@@ -0,0 +1,390 @@
+    virtual void visit(AstModule* nodep) VL_OVERRIDE {
+        // Don't visit twice
+        if (m_checked.find(nodep) != m_checked.end()) return;
+        UINFO(5, "Checking " << nodep->prettyNameQ() << " from "
+                             << (m_modp ? m_modp->prettyNameQ() : string("null")) << std::endl);
+        AstModule* const prevModulep = m_modp;
+        ModuleSet prevReferred;
+        V3HierBlock::GParams prevGParams;
+        if (nodep->hierBlock()) {
+            m_modp = nodep;
+            prevReferred.swap(m_referred);
+        }
+        prevGParams.swap(m_gparams);
+
+        iterateChildren(nodep);
+
+        if (nodep->hierBlock()) {
+            m_planp->add(nodep, m_gparams);
+            for (ModuleSet::const_iterator it = m_referred.begin(); it != m_referred.end(); ++it) {
+                m_planp->registerUsage(nodep, *it);
+            }
+            m_modp = prevModulep;
+            m_referred.swap(prevReferred);
+        }
+        m_gparams.swap(prevGParams);
+        m_checked.insert(nodep);
+    }
+    virtual void visit(AstCell* nodep) VL_OVERRIDE {
+        // Visit used module here to know that the module is hier_block or not.
+        // This visitor behaves almost depth first search
+        if (AstModule* modp = VN_CAST(nodep->modp(), Module)) {
+            iterate(modp);
+            m_referred.insert(modp);
+        }
+    }
+    virtual void visit(AstVar* nodep) VL_OVERRIDE {
+        if (nodep->isGParam() && nodep->overriddenParam()) m_gparams.push_back(nodep);
+    }
+
+public:
+    Visitor(V3HierBlockPlan* planp, AstNetlist* netlist)
+        : m_planp(planp)
+        , m_modp(NULL) {
+
+        iterateChildren(netlist);
+    }
+    VL_DEBUG_FUNC;  // Declare debug()
+};
+
+//######################################################################
+
+V3HierBlockPlan::V3HierBlockPlan() {}
+
+bool V3HierBlockPlan::isHierBlock(const AstNodeModule* modp) const {
+    return m_blocks.find(modp) != m_blocks.end();
+}
+
+void V3HierBlockPlan::add(const AstNodeModule* modp, const std::vector<AstVar*>& gparams) {
+    iterator it = m_blocks.find(modp);
+    if (it == m_blocks.end()) {
+        V3HierBlock* hblockp = new V3HierBlock(modp, gparams);
+        UINFO(3, "Add " << modp->prettyNameQ() << " with " << gparams.size() << " parameters"
+                        << std::endl);
+        m_blocks.insert(std::make_pair(modp, hblockp));
+    }
+}
+
+void V3HierBlockPlan::registerUsage(const AstNodeModule* parentp, const AstNodeModule* childp) {
+    iterator parent = m_blocks.find(parentp);
+    UASSERT_OBJ(parent != m_blocks.end(), parentp, "must be added");
+    iterator child = m_blocks.find(childp);
+    if (child != m_blocks.end()) {
+        UINFO(3, "Found usage relation " << parentp->prettyNameQ() << " uses "
+                                         << childp->prettyNameQ() << std::endl);
+        parent->second->addChild(child->second);
+        child->second->addParent(parent->second);
```

Comment:
I suppose this is OK, but FYI other places use a graph to represent this (e.g. LinkCells).

Reply:
Thank you for letting me know. I will try using a graph in the future PR if it will do a similar operation.